File: score_analyser.rb
require_relative "board_analyser"

class ScoreAnalyser

  def initialize
    @goban = nil
    @analyser = BoardAnalyser.new
  end

  # Compute simple score difference for a AI-AI game (score info not needed)
  def compute_score_diff(goban, komi)
    @analyser.count_score(goban)
    scores = @analyser.scores
    prisoners = @analyser.prisoners
    b = scores[BLACK] + prisoners[WHITE]
    w = scores[WHITE] + prisoners[BLACK] + komi
    return b - w
  end
  
  # Returns score info as an array of strings
  def compute_score(goban, komi, who_resigned)
    start_scoring(goban, komi, who_resigned)
    txt = score_info_to_s(@score_info)
    return txt
  end

  # Initialize scoring phase
  def start_scoring(goban, komi, who_resigned)
    @goban = goban
    if who_resigned
      winner = Grid::COLOR_NAMES[1-who_resigned]
      other = Grid::COLOR_NAMES[who_resigned]
      @score_info = "#{winner} won (since #{other} resigned)"
      return
    end
    @analyser.count_score(goban)
    scores = @analyser.scores
    prisoners = @analyser.prisoners
    totals = []
    details = []
    add_pris = true
    2.times do |c|
      kom = (c == WHITE ? komi : 0)
      pris = (add_pris ? prisoners[1 - c] : -prisoners[c])
      totals[c] = scores[c] + pris + kom
      details[c] = [scores[c], pris, kom]
    end
    @score_info = [totals, details]
  end
  
  def get_score
    return score_info_to_s(@score_info)
  end

  def score_info_to_s(info)
    return [info] if info.is_a?(String) # for games where all but 1 resigned
    raise "Invalid score info: #{info}" if !info or info.size!=2
    totals = info[0]
    details = info[1]
    raise "Invalid score info" if totals.size!=details.size
    s = []
    s.push(score_winner_to_s(totals))
    2.times do |c|
      detail = details[c]
      if detail == nil
        s.push("#{Grid.color_name(c)} resigned")
        next
      end
      raise "Invalid score details" if detail.size!=3
      score = detail[0]
      pris = detail[1]
      komi = detail[2]
      komi_str = (komi > 0 ? " + #{komi} komi" : "")
      s.push("#{Grid.color_name(c)} (#{Grid::COLOR_CHARS[c]}): "+
          "#{pts(totals[c])} (#{score} #{pris<0 ? '-' : '+'} #{pris.abs} prisoners#{komi_str})")
    end
    return s
  end
  
  def score_diff_to_s(diff)
    if diff != 0
      win = if diff > 0 then BLACK else WHITE end
      return "#{Grid.color_name(win)} wins by #{pts(diff.abs)}"
    else
      return "Tie game"
    end
  end
  
  def score_winner_to_s(totals)
    if totals.size == 2
      diff = totals[0] - totals[1]
      return score_diff_to_s(diff)
    else
      max = totals.max
      winners = []
      totals.size.times { |c| winners.push(c) if totals[c]==max }
      if winners.size == 1
        return "#{Grid.color_name(winners[0])} wins with #{pts(max)}"
      else
        return "Tie between " +
          winners.map{|w| "#{Grid.color_name(w)}"}.join(" & ") +
          ", #{winners.size==2 ? 'both' : 'all'} with #{pts(max)}"
      end
    end
  end

private
  
  def pts(n)
    return n!=1 ? "#{n} points" : "1 point"
  end

end

File: js/ScoreAnalyser.js
'use strict';

var CONST = require('./constants');
var main = require('./main');
var Grid = require('./Grid');

var BLACK = CONST.BLACK, WHITE = CONST.WHITE;


/** @class */
function ScoreAnalyser(game) {
    this.game = game;
    this.analyser = new main.defaultAi.BoardAnalyser(game);
}
module.exports = ScoreAnalyser;


// Computes score and returns it as a number >0 if black won
ScoreAnalyser.prototype.computeScoreDiff = function () {
    var totals = this._computeScore()[0];
    return totals[BLACK] - totals[WHITE];
};

// Computes score and returns it as an array of human-readable strings
 ScoreAnalyser.prototype.computeScoreAsTexts = function () {
    return this._scoreInfoToS(this._computeScore());
};

ScoreAnalyser.prototype.computeScoreInfo = function () {
    return this._computeScore();
};

// Retrives the scoring grid once the score has been computed
ScoreAnalyser.prototype.getScoringGrid = function () {
    return this.analyser.getScoringGrid();
};

ScoreAnalyser.prototype._computeScore = function () {
    var game = this.game;
    var goban = game.goban;
    var komi = game.komi;

    if (game.whoResigned !== null) {
        return ['resi', 1 - game.whoResigned, game.resignReason];
    }

    this.analyser.countScore(goban);

    var scores = this.analyser.scores;
    var prisoners = this.analyser.prisoners;

    var totals = [
        scores[BLACK] + prisoners[WHITE],
        scores[WHITE] + prisoners[BLACK] + komi];
    var details = [
        [scores[BLACK], prisoners[WHITE], 0],
        [scores[WHITE], prisoners[BLACK], komi]];

    return [totals, details];
};

function pointsToString(n) {
    return ( n !== 1 ? n + ' points' : '1 point' );
}

ScoreAnalyser.prototype._scoreInfoToS = function (info) {
    if (info[0] === 'resi') {
        var winner = Grid.COLOR_NAMES[info[1]], loser = Grid.COLOR_NAMES[1 - info[1]];
        var reason;
        switch (info[2]) {
        case 'time': reason = loser + ' ran out of time'; break;
        case null: case undefined: reason = loser + ' resigned'; break;
        default: reason = loser + ' disqualified: ' + info[2];
        }
        return [winner + ' won (' + reason + ')'];
    }
    if (!info || info.length !== 2) throw new Error('Invalid score info: ' + info);
    var totals = info[0];
    var details = info[1];
    if (totals.length !== details.length) throw new Error('Invalid score info');

    var s = [];
    var diff = totals[BLACK] - totals[WHITE];
    s.push(this._scoreDiffToS(diff));

    for (var c = BLACK; c <= WHITE; c++) {
        var detail = details[c];
        var score = detail[0], pris = detail[1], komi = detail[2];
        var prisoners = pris !== 0 ? ' ' + ( pris < 0 ? '-' : '+' ) + ' ' + Math.abs(pris) + ' prisoners' : '';
        var komiStr = (( komi > 0 ? ' + ' + komi + ' komi' : '' ));
        var detailStr = pris || komi ? ' (' + score + prisoners + komiStr + ')' : '';
        s.push(Grid.colorName(c) + ': ' + pointsToString(totals[c]) + detailStr);
    }
    return s;
};

ScoreAnalyser.prototype._scoreDiffToS = function (diff) {
    if (diff === 0) return 'Tie game';
    var win = ( diff > 0 ? BLACK : WHITE );
    return Grid.colorName(win) + ' wins by ' + pointsToString(Math.abs(diff));
};
